File: src/main/java/frc/robot/subsystems/Intake.java
// Copyright (c) 2022 FRC Team 2881 - The Lady Cans
//
// Open Source Software; you can modify and/or share it under the terms of BSD
// license file in the root directory of this project.

package frc.robot.subsystems;

import static frc.robot.Constants.Intake.kCurrentLimit;
import static frc.robot.Constants.Intake.kMaxSpeed;
import static frc.robot.Constants.Intake.kMotor;
import static frc.robot.Constants.Intake.kSolenoid;

import com.ctre.phoenix.motorcontrol.NeutralMode;
import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;

import edu.wpi.first.wpilibj.PneumaticsModuleType;
import edu.wpi.first.wpilibj.Solenoid;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

public class Intake extends SubsystemBase {
  private final WPI_TalonSRX m_intake = new WPI_TalonSRX(kMotor);
  private final Solenoid m_solenoid = new Solenoid(PneumaticsModuleType.CTREPCM, kSolenoid);
  private boolean shootExtend = false;
  private boolean buttonExtend = false;

  /** Creates a new Intake. */
  public Intake() {
    m_intake.setNeutralMode(NeutralMode.Brake);
    m_intake.configContinuousCurrentLimit(kCurrentLimit);
    m_intake.configPeakCurrentLimit(0);
    m_intake.enableCurrentLimit(true);
  }

  @Override
  public void periodic() {
    // This method will be called once per scheduler run
  }

  public void reset() {
    run(0.0);
    shootExtend = false;
    buttonExtend = false;
    retract();
  }

  public void run(double speed) {
    if(speed > kMaxSpeed) {
      speed = kMaxSpeed;
    }
    m_intake.set(-speed);
  }

  public void runReverse(double speed) {
    if(speed > kMaxSpeed) {
      speed = kMaxSpeed;
    }
    m_intake.set(speed);
  }

  public void extend() {
    buttonExtend = true;
    m_solenoid.set(true);
  }

  public void retract() {
    buttonExtend = false;
    if(shootExtend == false) {
      m_solenoid.set(false);
    }
  }

  public void shootExtend() {
    shootExtend = true;
    m_solenoid.set(true);
  }

  public void shootRetract() {
    shootExtend = false;
    if(buttonExtend == false) {
      m_solenoid.set(false);
    }
  }
}

File: core/subsystems/intake.py
from commands2 import Subsystem, Command
from wpilib import Compressor, Solenoid, PneumaticsModuleType
from phoenix5 import WPI_TalonSRX, NeutralMode
from lib import logger, utils
import core.constants as constants

class Intake(Subsystem):
  def __init__(self) -> None:
    super().__init__()
    self._constants = constants.Subsystems.Intake

    self._rollers = WPI_TalonSRX(5)
    self._rollers.setNeutralMode(NeutralMode.Brake)
    self._rollers.configContinuousCurrentLimit(30)
    self._rollers.configPeakCurrentLimit(0)
    self._rollers.enableCurrentLimit(True)

    self._compressor = Compressor(PneumaticsModuleType.CTREPCM)
    self._compressor.enableDigital()

    self._solenoid = Solenoid(PneumaticsModuleType.CTREPCM, 0)

  def periodic(self) -> None:
    self._updateTelemetry()

  def run_(self) -> Command:
    return self.startEnd(
      lambda: self._run(),
      lambda: self.reset()
    )

  def retract(self) -> Command:
    return self.runOnce(
      lambda: self._retract()
    )

  def _run(self) -> None:
    self._extend()
    self._rollers.set(0.75)

  def _extend(self) -> None:
    self._solenoid.set(True)

  def _retract(self) -> None:
    self._solenoid.set(False)
    self._rollers.set(0)

  def isExtended(self) -> bool:
    return self._solenoid.get()

  def reset(self) -> None:
    self._rollers.stopMotor()

  def _updateTelemetry(self) -> None:
    pass